core/src/main/java/org/kiwix/kiwixmobile/core/main/KiwixTextToSpeech.kt

```diff
@@ -0,0 +1,295 @@
+          context.toast(R.string.tts_lang_not_supported, Toast.LENGTH_LONG)
+          return
+        }
+        if (requestAudioFocus()) {
+          loadURL(webView)
+        }
+      }
+    }
+  }
+
+  private fun isMissingOrUnsupportedLanguage(languageAvailabilityResult: Int): Boolean =
+    languageAvailabilityResult == LANG_MISSING_DATA ||
+      languageAvailabilityResult == LANG_NOT_SUPPORTED
+
+  private fun getFeatures(
+    tts: TextToSpeech?
+  ): Set<String> = tts!!.voice.features
+
+  private fun loadURL(webView: WebView) {
+    // We use JavaScript to get the content of the page conveniently, earlier making some
+    // changes in the page
+    webView.loadUrl(
+      """
+        javascript:body = document.getElementsByTagName('body')[0].cloneNode(true);
+        // Remove some elements that are shouldn't be read (table of contents,
+        // references numbers, thumbnail captions, duplicated title, etc.)
+        "toRemove = body.querySelectorAll('sup.reference, #toc, .thumbcaption, 
+        "    title, .navbox');
+        "Array.prototype.forEach.call(toRemove, function(elem) {
+        "    elem.parentElement.removeChild(elem);
+        "});
+        "tts.speakAloud(body.innerText);
+        """
+    )
+  }
+
+  fun stop() {
+    tts.let {
+      {
+        if (it.stop() == SUCCESS) {
+          currentTTSTask = null
+          it.setOnUtteranceProgressListener(null)
+          onSpeakingListener.onSpeakingEnded()
+        }
+      }
+    }
+  }
+
+  private fun requestAudioFocus(): Boolean {
+    val audioFocusRequest = am.requestAudioFocus(
+      onAudioFocusChangeListener, AudioManager.STREAM_MUSIC,
+      AudioManager.AUDIOFOCUS_GAIN
+    )
+    Log.d(TAG_KIWIX, "Audio Focus Requested")
+    synchronized(focusLock) {
+      return@requestAudioFocus audioFocusRequest ==
+        AudioManager.AUDIOFOCUS_REQUEST_GRANTED
+    }
+  }
+
+  fun pauseOrResume() {
+    currentTTSTask?.let {
+      if (it.paused) {
+        if (!requestAudioFocus()) return@pauseOrResume
+      } else {
+        it.pause()
+      }
+    }
+  }
+
+  fun initWebView(webView: WebView) {
+    webView.addJavascriptInterface(TTSJavaScriptInterface(), "tts")
+  }
+
+  /**
+   * Releases the resources used by the engine.
+   *
+   * @see android.speech.tts.TextToSpeech.shutdown
+   */
+  fun shutdown() {
+    tts.shutdown()
+  }
+
+  /**
+   * The listener which is notified when initialization of the TextToSpeech engine is successfully
+   * done.
+   */
+  internal interface OnInitSucceedListener {
+    fun onInitSucceed()
+  }
+
+  /**
+   * The listener that is notified when speaking starts or stops (regardless of whether it was a
+   * result of error, user, or because whole text was read).
+   *
+   *
+   * Note that the methods of this interface may not be called from the UI thread.
+   */
+  interface OnSpeakingListener {
+    fun onSpeakingStarted()
+    fun onSpeakingEnded()
+  }
+
+  inner class TTSTask
```

Comment:
this looks odd, to me looks like it should fit in our line length, press `ctrl+shift+j` with your cursor on this line to join the lines